httpx/dispatch/http2.py

```diff
@@ -55,16 +56,23 @@ def init_complete(self) -> BaseEvent:
     async def send(self, request: Request, timeout: Timeout = None) -> Response:
         timeout = Timeout() if timeout is None else timeout
 
-        if not self.init_started:
-            # The very first stream is responsible for initiating the connection.
-            self.init_started = True
-            await self.send_connection_init(timeout)
-            stream_id = self.state.get_next_available_stream_id()
-            self.init_complete.set()
-        else:
-            # All other streams need to wait until the connection is established.
-            await self.init_complete.wait()
-            stream_id = self.state.get_next_available_stream_id()
+        try:
+            if not self.init_started:
+                # The very first stream is responsible
+                # for initiating the connection.
+                self.init_started = True
+                await self.send_connection_init(timeout)
+                stream_id = self.state.get_next_available_stream_id()
+                self.init_complete.set()
+            else:
+                # All other streams need to wait until the
+                # connection is established.
+                await self.init_complete.wait()
+                stream_id = self.state.get_next_available_stream_id()
+        except NoAvailableStreamIDError:
+            self.state = h2.connection.H2Connection(config=self.CONFIG)
```

Comment:
We can't do this, since it'll break existing streams.

Getting this right is a little bit complex, and I think we should probably pause on it, pending some more comprehensive work around connection state management along the lines in https://github.com/encode/httpx/pull/624